File: e2e/go/error_test.go
// E2e tests for category: error
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_Error401Unauthorized(t *testing.T) {
	// Handles 401 Unauthorized response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error403Forbidden(t *testing.T) {
	// Handles 403 Forbidden response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error404Page(t *testing.T) {
	// Handles 404 response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error408RequestTimeout(t *testing.T) {
	// Handles 408 Request Timeout response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error410Gone(t *testing.T) {
	// Handles 410 Gone response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error500Server(t *testing.T) {
	// Handles 500 server error
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_Error502BadGateway(t *testing.T) {
	// Handles 502 Bad Gateway response correctly
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorConnectionRefused(t *testing.T) {
	// Handles connection refused error gracefully
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorDnsResolution(t *testing.T) {
	// Handles DNS resolution failure gracefully
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorEmptyResponse(t *testing.T) {
	// Handles 200 with completely empty body gracefully
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.HtmlNotEmpty != false {
		t.Errorf("equals mismatch: got %q", result.HtmlNotEmpty)
	}
	if result.Error.IsError != false {
		t.Errorf("equals mismatch: got %q", result.Error.IsError)
	}
}

func Test_ErrorInvalidProxy(t *testing.T) {
	// Proxy pointing to unreachable address causes connection error during scrape
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorPartialResponse(t *testing.T) {
	// Handles incomplete or truncated HTTP response
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorRateLimited(t *testing.T) {
	// Handles 429 rate limiting with Retry-After
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorRetry503(t *testing.T) {
	// Retries request on 503 Service Unavailable response
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorRetryBackoff(t *testing.T) {
	// Implements exponential backoff when retrying failed requests
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorSslInvalidCert(t *testing.T) {
	// Handles SSL certificate validation error
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorTimeout(t *testing.T) {
	// Handles request timeout
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorWafAkamai(t *testing.T) {
	// Akamai WAF detection returns WafBlocked error
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorWafFalse403(t *testing.T) {
	// Detects WAF/bot protection false 403 (Cloudflare challenge page)
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

func Test_ErrorWafImperva(t *testing.T) {
	// Imperva/Incapsula WAF detection
	_, err := pkg.Scrape()
	if err == nil {
		t.Errorf("expected an error, but call succeeded")
	}
}

File: e2e/java/src/test/java/dev/kreuzberg/e2e/ErrorTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/** E2e tests for category: error. */
class ErrorTest {
    @Test
    void testError401Unauthorized() {
        // Handles 401 Unauthorized response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError403Forbidden() {
        // Handles 403 Forbidden response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError404Page() {
        // Handles 404 response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError408RequestTimeout() {
        // Handles 408 Request Timeout response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError410Gone() {
        // Handles 410 Gone response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError500Server() {
        // Handles 500 server error
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testError502BadGateway() {
        // Handles 502 Bad Gateway response correctly
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorConnectionRefused() {
        // Handles connection refused error gracefully
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorDnsResolution() {
        // Handles DNS resolution failure gracefully
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorEmptyResponse() {
        // Handles 200 with completely empty body gracefully
        var result = Kreuzcrawl.scrape();
        assertEquals(false, result.html_not_empty());
        assertEquals(false, result.error().is_error());
    }

    @Test
    void testErrorInvalidProxy() {
        // Proxy pointing to unreachable address causes connection error during scrape
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorPartialResponse() {
        // Handles incomplete or truncated HTTP response
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorRateLimited() {
        // Handles 429 rate limiting with Retry-After
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorRetry503() {
        // Retries request on 503 Service Unavailable response
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorRetryBackoff() {
        // Implements exponential backoff when retrying failed requests
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorSslInvalidCert() {
        // Handles SSL certificate validation error
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorTimeout() {
        // Handles request timeout
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorWafAkamai() {
        // Akamai WAF detection returns WafBlocked error
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorWafFalse403() {
        // Detects WAF/bot protection false 403 (Cloudflare challenge page)
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

    @Test
    void testErrorWafImperva() {
        // Imperva/Incapsula WAF detection
        assertThrows(Exception.class, () -> Kreuzcrawl.scrape());
    }

}
